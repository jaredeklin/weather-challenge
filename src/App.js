import React, { Component } from 'react';
import './App.css';
import Search from './Search';
import { Api } from './Api';
import { Current } from './Current';
import { ForecastContainer } from './ForecastContainer';
import { Error } from './Error';

class App extends Component {
  constructor() {
    super();
    this.state = {
      currentLocation: '',
      weatherData: {},
      error: false,
      loading: false
    };

    this.api = new Api();
  }

  getLocationData = async location => {
    const currentData = await this.api.getCurrent(location);
    const forecastData = await this.api.getForecast(currentData.id);
    console.log(currentData, forecastData);

    const weatherData = { ...currentData, forecast: forecastData };

    this.setState({
      currentLocation: currentData.name,
      weatherData
    });
  };

  render() {
    const { currentLocation, weatherData, error } = this.state;
    return (
      <div className="App">
        <Search getLocationData={this.getLocationData} />
        {error && <Error />}
        {!error && (
          <div>
            {currentLocation === '' && (
              <h1>Hello! Enter a location to view the weather</h1>
            )}
            {currentLocation !== '' && (
              <section>
                <Current weatherData={weatherData} />
                <ForecastContainer weatherData={weatherData.forecast} />
              </section>
            )}
          </div>
        )}
      </div>
    );
  }
}

export default App;
