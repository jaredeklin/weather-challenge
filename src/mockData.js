export const mockCurrentData = {
  base: 'stations',
  clouds: { all: 1 },
  cod: 200,
  coord: { lon: -104.98, lat: 39.74 },
  dt: 1540593360,
  id: 5419384,
  main: {
    temp: 66.18,
    pressure: 1019,
    humidity: 14,
    temp_min: 55.4,
    temp_max: 71.6
  },
  name: 'Denver',
  sys: {
    type: 1,
    id: 602,
    message: 0.0049,
    country: 'US',
    sunrise: 1540560180
  },
  visibility: 16093,
  weather: [
    {
      description: 'clear sky',
      icon: '01d',
      id: 800,
      main: 'Clear'
    }
  ],
  wind: { speed: 5.82, deg: 50 }
};

export const mockForecastData = {
  city: {
    coord: { lat: 39.7392, lon: -104.9848 },
    country: 'US',
    id: 5419384,
    name: 'Denver'
  },
  cnt: 40,
  cod: '200',
  list: [
    {
      clouds: { all: 0 },
      dt: 1540598400,
      dt_txt: '2018-10-27 00:00:00',
      main: {
        temp: 59.97,
        temp_min: 54.85,
        temp_max: 59.97,
        pressure: 818.16,
        sea_level: 1026.38
      },
      sys: { pod: 'n' },
      weather: [
        {
          description: 'clear sky',
          icon: '01n',
          id: 800,
          main: 'Clear'
        }
      ],
      wind: { speed: 8.77, deg: 314.503 }
    },
    {
      clouds: { all: 0 },
      dt: 1540598400,
      dt_txt: '2018-10-27 18:00:00',
      main: {
        temp: 59.97,
        temp_min: 54.85,
        temp_max: 59.97,
        pressure: 818.16,
        sea_level: 1026.38
      },
      sys: { pod: 'n' },
      weather: [
        {
          description: 'clear sky',
          icon: '01n',
          id: 800,
          main: 'Clear'
        }
      ],
      wind: { speed: 8.77, deg: 314.503 }
    },
    {
      clouds: { all: 0 },
      dt: 1540598400,
      dt_txt: '2018-10-27 18:00:00',
      main: {
        temp: 59.97,
        temp_min: 54.85,
        temp_max: 59.97,
        pressure: 818.16,
        sea_level: 1026.38
      },
      sys: { pod: 'n' },
      weather: [
        {
          description: 'clear sky',
          icon: '01n',
          id: 800,
          main: 'Clear'
        }
      ],
      wind: { speed: 8.77, deg: 314.503 }
    },
    {
      clouds: { all: 0 },
      dt: 1540598400,
      dt_txt: '2018-10-27 18:00:00',
      main: {
        temp: 59.97,
        temp_min: 54.85,
        temp_max: 59.97,
        pressure: 818.16,
        sea_level: 1026.38
      },
      sys: { pod: 'n' },
      weather: [
        {
          description: 'clear sky',
          icon: '01n',
          id: 800,
          main: 'Clear'
        }
      ],
      wind: { speed: 8.77, deg: 314.503 }
    },
    {
      clouds: { all: 0 },
      dt: 1540598400,
      dt_txt: '2018-10-27 18:00:00',
      main: {
        temp: 59.97,
        temp_min: 54.85,
        temp_max: 59.97,
        pressure: 818.16,
        sea_level: 1026.38
      },
      sys: { pod: 'n' },
      weather: [
        {
          description: 'clear sky',
          icon: '01n',
          id: 800,
          main: 'Clear'
        }
      ],
      wind: { speed: 8.77, deg: 314.503 }
    },
    {
      clouds: { all: 0 },
      dt: 1540598400,
      dt_txt: '2018-10-27 18:00:00',
      main: {
        temp: 59.97,
        temp_min: 54.85,
        temp_max: 59.97,
        pressure: 818.16,
        sea_level: 1026.38
      },
      sys: { pod: 'n' },
      weather: [
        {
          description: 'clear sky',
          icon: '01n',
          id: 800,
          main: 'Clear'
        }
      ],
      wind: { speed: 8.77, deg: 314.503 }
    },
    {
      clouds: { all: 0 },
      dt: 1540598400,
      dt_txt: '2018-10-27 15:00:00',
      main: {
        temp: 59.97,
        temp_min: 54.85,
        temp_max: 59.97,
        pressure: 818.16,
        sea_level: 1026.38
      },
      sys: { pod: 'n' },
      weather: [
        {
          description: 'clear sky',
          icon: '01n',
          id: 800,
          main: 'Clear'
        }
      ],
      wind: { speed: 8.77, deg: 314.503 }
    }
  ]
};

export const mockExpectedForecastData = [
  {
    date: 'Sat',
    description: 'Clear',
    imgUrl: 'http://openweathermap.org/img/w/01n.png',
    temp: '60'
  },
  {
    date: 'Sat',
    description: 'Clear',
    imgUrl: 'http://openweathermap.org/img/w/01n.png',
    temp: '60'
  },
  {
    date: 'Sat',
    description: 'Clear',
    imgUrl: 'http://openweathermap.org/img/w/01n.png',
    temp: '60'
  },
  {
    date: 'Sat',
    description: 'Clear',
    imgUrl: 'http://openweathermap.org/img/w/01n.png',
    temp: '60'
  },
  {
    date: 'Sat',
    description: 'Clear',
    imgUrl: 'http://openweathermap.org/img/w/01n.png',
    temp: '60'
  }
];

export const mockExpectedCurrentData = {
  date: 'Saturday, December 1st',
  description: 'Clear',
  forecast: [
    {
      date: 'Sat',
      description: 'Clear',
      imgUrl: 'http://openweathermap.org/img/w/01n.png',
      temp: '60'
    },
    {
      date: 'Sat',
      description: 'Clear',
      imgUrl: 'http://openweathermap.org/img/w/01n.png',
      temp: '60'
    },
    {
      date: 'Sat',
      description: 'Clear',
      imgUrl: 'http://openweathermap.org/img/w/01n.png',
      temp: '60'
    },
    {
      date: 'Sat',
      description: 'Clear',
      imgUrl: 'http://openweathermap.org/img/w/01n.png',
      temp: '60'
    },
    {
      date: 'Sat',
      description: 'Clear',
      imgUrl: 'http://openweathermap.org/img/w/01n.png',
      temp: '60'
    }
  ],
  id: 5419384,
  imgUrl: 'http://openweathermap.org/img/w/01d.png',
  name: 'Denver',
  temp: '66'
};
