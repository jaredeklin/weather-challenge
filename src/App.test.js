import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';
import { mockExpectedCurrentData } from './mockData';

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('getLocationData', () => {
    const mockLocation = 'Denver';

    it('should call api.getCurrent with correct params', async () => {
      const mockGetCurrent = (wrapper.instance().api.getCurrent = jest.fn());

      wrapper.instance().getLocationData(mockLocation);
      expect(mockGetCurrent).toHaveBeenCalledWith(mockLocation);
    });

    it('shpuld update error state if getCurrent returns with an error', async () => {
      wrapper.instance().api.getCurrent = jest.fn().mockReturnValue(null);

      await wrapper.instance().getLocationData(mockLocation);
      expect(wrapper.state('error')).toEqual(true);
    });

    it('should setState with correct weather data', async () => {
      const expectedState = {
        currentLocation: 'Denver',
        weatherData: mockExpectedCurrentData,
        error: false
      };

      wrapper.instance().api.getCurrent = jest
        .fn()
        .mockReturnValue(mockExpectedCurrentData);

      await wrapper.instance().getLocationData(mockLocation);
      expect(wrapper.state()).toEqual(expectedState);
    });
  });
});
