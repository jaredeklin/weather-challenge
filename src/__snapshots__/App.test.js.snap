// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Search
          getLocationData={[Function]}
        />,
        false,
        <div>
          <h1>
            Hello! Enter a location to view the weather
          </h1>
        </div>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "getLocationData": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              Hello! Enter a location to view the weather
            </h1>,
            false,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Hello! Enter a location to view the weather",
            },
            "ref": null,
            "rendered": "Hello! Enter a location to view the weather",
            "type": "h1",
          },
          false,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Search
            getLocationData={[Function]}
          />,
          false,
          <div>
            <h1>
              Hello! Enter a location to view the weather
            </h1>
          </div>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "getLocationData": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                Hello! Enter a location to view the weather
              </h1>,
              false,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Hello! Enter a location to view the weather",
              },
              "ref": null,
              "rendered": "Hello! Enter a location to view the weather",
              "type": "h1",
            },
            false,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
